### PR TITLE
Generic regridding preprocessor

### PR DESCRIPTION
**Pull request by zklaus**
## Description
This adds the long-promised iris-esmf-regrid based regridding to the regrid preprocessor.
It does so in a generic way, meaning the preprocessor is not actually adapted to the iris-esmf-regrid package, but rather gets the ability to load any kind of Iris regridding scheme by specifying the corresponding factory, also allowing for the passing of arbitrary arguments.

I'd like to solicit some early feedback!

Do you think this is feasible, i.e. using a generic loading mechanism?

If you want to try it out, don't forget to create a new environment that contains the iris-esmf-regrid conda package as mentioned in the environment.yml in this PR (of course you can also just install it in your environment).
You can use the corresponding branch iris-esmf-regrid and the `recipe_ocean_multimap.yml` in it to try things out.

Documentation:
- Preprocessor docs: https://esmvaltool--1448.org.readthedocs.build/projects/ESMValCore/en/1448/recipe/preprocessor.html#regridding-interpolation-extrapolation-schemes
- API docs: https://esmvaltool--1448.org.readthedocs.build/projects/ESMValCore/en/1448/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.regrid

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by schlunma** [comment]
I just successfully tested this PR with a couple if `iris` schemes (e.g. `iris.analysis.Linear`) :tada: It works really well and just like expected.

Three small comments:
- If an invalid scheme is provided, the error message can be really cryptic. Might be useful to catch the original exception and use a `raise from` with a more user-friendly message.
- Similarly, if `scheme` is not given in the `dict`, this leads to a `KeyError: 'scheme'`. This message should be improved.
- I think we should rename the `scheme` key in the `dict` to something else to avoid duplicating the word "scheme". Maybe something like `name`, `module`, or `method`?

```yml
  regrid:
    regrid:
      target_grid: 0.5x0.5
      scheme:
        name/module/method: iris.analysis.Linear
        extrapolation_mode: mask
```

File: esmvalcore/preprocessor/_regrid.py [hunk 1]
```diff
@@ -503,7 +504,14 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
     --------
     extract_levels : Perform vertical regridding.
     """
-    if HORIZONTAL_SCHEMES.get(scheme.lower()) is None:
+    if isinstance(scheme, dict):
+        module_name, scheme_name = scheme.pop("scheme").rsplit(".", 1)
+        scheme_module = importlib.import_module(module_name)
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_regrid.py:509]: I would sneak in an `ImportError` exception catcher here

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
nice idea, I likes! One quick prelim suggestion from me, and of course, as Lenin used to say - tests, tests and tests again :grin: (lemme know I can help with the tests)

**Comment by zklaus** [comment]
Alright, thanks @schlunma and @valeriupredoi. I'll take that as an initial go-ahead and will come back with a more fleshed out version.

**Comment by schlunma** [comment]
@zklaus You think this will be ready in time for v2.5.0? I think it would be nice to include this.

**Comment by zklaus** [comment]
Now on to writing some tests.

**Comment by zklaus** [comment]
The main purpose of this PR is to allow the use of iris-esmf-regrid to enable faster, lazy regridding all around and to facilitate regridding of unstructured grids with the new iris mesh support. However, the mechanism introduced here is more general, and hence I introduce neither iris-esmf-regrid as a dependency, nor extensive specific documentation. For documentation we rely on the documentation of iris-esmf-regrid itself; as a dependency this should be introduced alongside the recipe that makes use of it.

Of course, this can be debated. If you prefer to integrate things more tightly here, we can take this now, or in a later PR for the next version. That would still allow us to do more experimentation with the released core in the next development cycle of the tool.

**Comment by schlunma** [comment]
Great, thanks Klaus!

I think it would have been nice to already include the `iris-esmf-regrid` dependency with this PR, but I'm also fine with just adding this framework. Please make sure to open an issue about this (with the label v2.6) so we don't forget about this.

Will have a look at this PR now!

**Comment by schlunma** [comment]
One further thought about this after reading the doc: I noticed that you use `iris-esmf-regrid` quite often in the doc. If `iris-esmf-regrid` is mentioned there, I think we need to include this.

Would it make sense to include it as a dependency, add a very small test, e.g.,

```py
def test_regrid_generic_regridding(self):
    regrid(self.src_cube, '1x1', {"reference": "esmf_regrid.schemes:ESMFAreaWeighted"})
```

and add a note to the doc that this feature is still "experimental"?

File: esmvalcore/preprocessor/_regrid.py [hunk 2]
```diff
@@ -478,13 +479,12 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
...
-        ``linear``,
-        ``linear_extrapolate``,
-        ``nearest``,
-        ``area_weighted``,
+    scheme : str or dict
+        The regridding scheme to perform. If both source and target grid are
+        structured (regular or irregular), can be one of the built-in schemes
+        ``linear``, ``linear_extrapolate``, ``nearest``, ``area_weighted``,
         ``unstructured_nearest``.
+        Alternatively, a `dict` that specifies generic regridding (see below).
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:487]:
```suggestion
        Alternatively, a ``dict`` that specifies generic regridding (see below).
```

File: esmvalcore/preprocessor/_regrid.py [hunk 3]
```diff
@@ -502,8 +502,71 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
     See Also
     --------
     extract_levels : Perform vertical regridding.
+
+    Notes
+    -----
+    This preprocessor allows for the use of arbitrary Iris regridding schemes,
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:508]:
```suggestion
    This preprocessor allows for the use of arbitrary :mod:`iris` regridding schemes,
```

File: esmvalcore/preprocessor/_regrid.py [hunk 4]
```diff
@@ -502,8 +502,71 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
...
+    that is anything that can be passed as a scheme to
+    :meth:`iris.cube.Cube.regrid` is possible. This enables the use of further
+    parameters for existing schemes, as well as the use of more advanced
+    schemes for example for unstructured meshes.
+    To use this functionality, a dictionary must be passed for the scheme with
+    a mandatory entry of `reference` in the form specified for the object
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:514]:
```suggestion
    a mandatory entry of ``reference`` in the form specified for the object
```

File: esmvalcore/preprocessor/_regrid.py [hunk 5]
```diff
@@ -502,8 +502,71 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
...
+    reference of the `entry point data model <https://packaging.python.org/en/
+    latest/specifications/entry-points/#data-model>`_,
+    i.e. `importable.module:object.attr`. This is used as a factory for the
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:517]:
```suggestion
    i.e. ``importable.module:object.attr``. This is used as a factory for the
```

File: esmvalcore/preprocessor/_regrid.py [hunk 6]
```diff
@@ -502,8 +502,71 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
...
+
+        my_preprocessor:
+          regrid:
+            target: 1x1
+            scheme:
+              reference: iris.analysis:Linear
+              extrapolation_mode: nanmask
+
+    To use the area weighted regridder available in
+    :class:`esmf_regrid.ESMFAreaWeighted`, make sure that `iris-esmf-regrid` is
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:534]:
```suggestion
    :class:`esmf_regrid.ESMFAreaWeighted`, make sure that ``iris-esmf-regrid`` is
```

**Review by schlunma — COMMENTED** [review]
A couple of minor comments on the doc. It would be great if you can also expand the doc [here](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/preprocessor.html#regridding-interpolation-extrapolation-schemes).

About the failing test: I just installed ESMValTool and ESMValCore in development mode on my machine without any problems. I also successfully ran a recipe and have a working version of `iris-esmf-regrid` (ran your test successfully). No idea what's going wrong in the CI machine...

File: esmvalcore/preprocessor/_regrid.py (hunk 2 above)
Comment (zklaus) [reply esmvalcore/preprocessor/_regrid.py:487]: Numpy style docstrings like we use should use single backticks here, I believe (see https://numpydoc.readthedocs.io/en/latest/format.html#common-rest-concepts).

File: esmvalcore/preprocessor/_regrid.py (hunk 3 above)
Comment (zklaus) [reply esmvalcore/preprocessor/_regrid.py:508]: Hm. It's not realy about the `iris` module, is it? How about
```suggestion
    This preprocessor allows for the use of arbitrary :doc:`iris:index` regridding schemes,
```

**Comment by zklaus** [comment]
I know what's wrong. There is no PyPI package for iris-esmf-regrid yet. Consequently, any attempt to install it directly via pip fails. I don't think that is much of a limitation in reality, but I will remove the package from `setup.py` for now and start the process of getting a PyPI package for iris-esmf-regrid up.

**Comment by zklaus** [comment]
Actually forget my last comment (that I have now hidden). The package is there. I'll investigate. I will say that I got a response of "currently the PyPI servers are overloaded" when trying `pip search esmf_regrid` locally before, but I am not completely sure that was the problem for the CI as well.

**Comment by valeriupredoi** [comment]
can't find it, man - use the web based search https://pypi.org/search/?q=iris-esmf-regrid

**Comment by valeriupredoi** [comment]
this one maybe? Says it's the bridge between Iris and ESMF (A Bridge Too Far, if you asked me :laughing: ) https://pypi.org/project/esmf-regrid/

**Comment by zklaus** [comment]
Yeah, of course that's the one. As named in setup.py. But what are you trying to say?

**Comment by valeriupredoi** [comment]
nothing, I basically just landed in this discussion and still need to figure my way around - maybe I should not comment until then :laughing:

File: esmvalcore/preprocessor/_regrid.py (hunk 3 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_regrid.py:508]: I'm not sure how this is rendered, so please just make the change. You can also ignore this comment if you like :smile:

Comment (zklaus) [reply esmvalcore/preprocessor/_regrid.py]: I made a change in the spirit and [here](https://esmvaltool--1448.org.readthedocs.build/projects/ESMValCore/en/1448/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.regrid) is how it renders. Let me know if you would like a further change.

**Comment by schlunma** [comment]
Ahh...well that's bad. I don't think there is an easy way to solve this apart from dropping support for python 3.7, right?

In that case it's probably best to remove it from the environment again (and also remove the test) and explicitly mention in the doc that it needs to be installed manually and requires python 3.8.

**Comment by valeriupredoi** [comment]
Klaus opened an issue here https://github.com/SciTools-incubator/iris-esmf-regrid/issues/161 - I've been snooping around (mostly looking at their [conda lock file](https://github.com/SciTools-incubator/iris-esmf-regrid/blob/main/requirements/nox.lock/py38-linux-64.lock)) and it's not obvious to me at all why Python<3.9 - all deps seem to be at their latest versions but built specifically against Python=3.8

**Comment by schlunma** [comment]
Ahh, it's python=3.8, not python>=3.8, misunderstood that :grimacing:

**Comment by valeriupredoi** [comment]
:snake: < :three: :nine: Manu

**Comment by zklaus** [comment]
I think in any case it is reasonable to remove it from the environment. After all, it really is only an example (from the point of view of the generic regridder).

**Comment by schlunma** [comment]
Agreed, but then we need to make it crystal clear in the documentation that `iris-esmf-regrid` needs to be installed manually and that only python<3.8 is supported.

**Comment by zklaus** [comment]
I have removed the dependency and improved the tests.

I am not so sure we need to be very forceful about how to install iris-esmf-regrid.

First, it really is only an example. If you want to use a different regridder or advanced features of the Iris regridders (perhaps some that will be added in the future), this works just as well, so I don't want to focus too much on this particular package.
Second, practically, it will not be much of an issue. Most people will not insist on a Python version and just ask for things to be installed, which will work out fine. We run into the problem in the CI because we test all versions.
Third, it is likely to change and since we also state that it is experimental, people should not be surprised with hiccups like this.

So imho it's better to have simple, accessible documentation and to refer to the iris-esmf-regrid documentation for further details.


Now on to writing the extended documentation requested by @schlunma.

**Comment by zklaus** [comment]
Done. One last round of reviews, @schlunma? @valeriupredoi do you want to have a look too and see if your blocking change request is sufficiently addressed?

**Comment by valeriupredoi** [comment]
cheers, Klaus! I'll uncork meself in a minute :+1:

File: doc/recipe/preprocessor.rst [hunk 7]
```diff
@@ -867,6 +877,67 @@ See also :func:`esmvalcore.preprocessor.regrid`
...
+          target_grid: 2.5x2.5
+          scheme:
+            reference: iris.analysis:AreaWeighted
+            mdtol: 0.7
+
+The value of the ``reference`` key has two parts that are separated by a
+``:`` with no surrounding spaces. The first part is an importable Python
+module, the second refers to the scheme, i.e. some callable that will be called
+with the remaining entries of the ``scheme`` dictionary passed as keyword
+arguments.
```
Comment (valeriupredoi) [inline doc/recipe/preprocessor.rst:919]:
```suggestion
argument.
```

File: esmvalcore/preprocessor/_regrid.py [hunk 8]
```diff
@@ -502,8 +502,71 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
...
+
+        my_preprocessor:
+          regrid:
+            target: 1x1
+            scheme:
+              reference: esmf_regrid:ESMFAreaWeighted
+
+    .. note::
+
+        Note that :doc:`iris-esmf-regrid:index` is still experimental.
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_regrid.py:547]: maybe add a TODO so we remove this when it's not experimental anymore?

**Review by valeriupredoi — APPROVED** [review]
very nice, great work, Klaus! Juat a couple typos from me, all good :beers:

File: esmvalcore/preprocessor/_regrid.py (hunk 8 above)
Comment (zklaus) [reply esmvalcore/preprocessor/_regrid.py:547]: Yeah, I was thinking the same thing. In fact, I had even typed it out and then removed it again. Do we have a standard way to note this kind of TODO? I think up to now we only have used issues for that purpose, no?

Comment (valeriupredoi) [reply esmvalcore/preprocessor/_regrid.py:547]: don't think we have any (only for deprecated stuff), maybe just plug one simple one and in 4 years we'll remove it if we don't get nuked in the meantime :grin:

File: doc/recipe/preprocessor.rst (hunk 7 above)
Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:919]: well, it's one dictionary = one keyword argument, but am not fussed, keep it as it is, me and grammatical correctness are not friends anyway

Comment (valeriupredoi) [reply doc/recipe/preprocessor.rst:919]: ah gotcha, cheers, man! All clear for take off :airplane:

**Review by schlunma — COMMENTED** [review]
Cheers Klaus, very nice work!

I tested this with actual data and it seems to work fine.

One very tiny comment on the doc.

File: esmvalcore/preprocessor/_regrid.py [hunk 9]
```diff
@@ -502,8 +502,71 @@ def regrid(cube, target_grid, scheme, lat_offset=True, lon_offset=True):
...
+
+        my_preprocessor:
+          regrid:
+            target: 1x1
+            scheme:
+              reference: iris.analysis:Linear
+              extrapolation_mode: nanmask
+
+    To use the area weighted regridder available in
+    :class:`esmf_regrid.ESMFAreaWeighted`, make sure that ``iris-esmf-regrid``
```
[the remark this answers is not in the document]
Comment (zklaus) [reply esmvalcore/preprocessor/_regrid.py]: Good catch. The reason is that the scheme really is in `esmf_regrid.schemes.ESMFAreaWeighted` but made available via some import into `esmf_regrid`. For now, I have made the module explicit; maybe we can simplify it again later when things have settled.

**Review by schlunma — APPROVED** [review]

**Pull request merged**